techtv2ovs/utils.py

```diff
@@ -190,6 +195,51 @@ def process_thumbs(self, ttv_video):
             ttv_video.thumbnail_status = ImportStatus.COMPLETE
         ttv_video.save()
 
+    def process_captions(self, ttv_video):
+        """
+        Copy over any caption files for a TechTVVideo from Dropbox to S3.
+        Create a VideoSubtitle object for each.
+
+        Args:
+            ttv_video (TechTVVideo): The TechTVVideo object to process
+        """
+        ttv_video.subtitle_status = ImportStatus.CREATED
+        captions = [caption[0] for caption in mysql_query(
+            'SELECT srt_file_file_name from closed_caption_files where video_id = %s', [ttv_video.ttv_id],
+            self.connection
+        )]
+        for caption in captions:
+            try:
+                basename, _ = splitext(caption)
+                dbx = dropbox.Dropbox(settings.DROPBOX_TOKEN)
+                _, response = dbx.files_download('/ODL-Video/TechTV/Captions/{}'.format(caption))
+                srt_bytes = BytesIO(response.content)
+                bucket = get_bucket(settings.VIDEO_S3_SUBTITLE_BUCKET)
+                bucket.upload_fileobj(
+                    Fileobj=srt_bytes,
+                    Key='subtitles/techtv/{}/{}'.format(ttv_video.video.hexkey, caption)
+                )
+                vtt_key = 'subtitles/techtv/{}/{}.vtt'.format(ttv_video.video.hexkey, basename)
+                converter = CaptionConverter()
+                converter.read(response.content.decode('utf-8'), SRTReader())
+                vtt_content = converter.write(WebVTTWriter())
+
+                bucket.upload_fileobj(
+                    Fileobj=BytesIO(vtt_content.encode('utf-8')),
+                    Key=vtt_key
+                )
+                VideoSubtitle.objects.get_or_create(s3_object_key=vtt_key, defaults={
+                    'video': ttv_video.video
+                })
+            except Exception as exc:
+                self.output.write("Error: caption file {} for ttv video {}".format(caption, ttv_video.ttv_id))
+                ttv_video.status = ImportStatus.ERROR
+                ttv_video.subtitle_status = ImportStatus.ERROR
+                ttv_video.errors += '{}\n\n'.format(str(exc))
```

Comment:
We might want to include the stack trace here too

Reply:
✅